File: t/average.c
/* t/average.c */

#include <math.h>
#include <stdio.h>
#include <string.h>

#include "average.h"
#include "framework.h"
#include "tap.h"

#define NOWT 0.0000001

static double xn(int n) {
  return (double)((100 - n) * 3.5 * ((n & 1) ? -1.3 : 3.2));
}

static void compute_average(double (*fn)(int n), int from, int to,
                            double *minp, double *avgp, double *maxp) {
  double total = 0;
  double min = NAN, max = NAN;
  for (int i = from; i <= to; i++) {
    double datum = fn(i);
    if (isnan(min) || datum < min) min = datum;
    if (isnan(max) || datum > max) max = datum;
    total += datum;
  }
  *minp = min;
  *avgp = total / (double)(to - from + 1);
  *maxp = max;
}

static void test_average(void) {
  for (unsigned len = 1; len < 5; len++) {
    nest_in("length: %u", len);
    average *avg = average_new(len);

    for (int x = 0; x < (int) len * 2; x++) {
      double datum = xn(x);
      nest_in("datum[%d]=%.4f", x, datum);

      double got_avg = average_push(avg, datum);
      double got_min = average_min(avg);
      double got_max = average_max(avg);

      int used = x + 1;
      if (used > (int) len) used = len;

      double want_min, want_avg, want_max;
      compute_average(xn, x - used + 1, x, &want_min, &want_avg, &want_max);

      if (!close_to(want_avg, got_avg, "average is %f", want_avg)) {
        diag("wanted %8.4f", want_avg);
        diag("   got %8.4f", got_avg);
      }

      if (!close_to(want_min, got_min, "min is %f", want_min)) {
        diag("wanted %8.4f", want_min);
        diag("   got %8.4f", got_min);
      }

      if (!close_to(want_max, got_max, "max is %f", want_max)) {
        diag("wanted %8.4f", want_max);
        diag("   got %8.4f", got_max);
      }

      ok(average_used(avg) == (unsigned) used, "used is %d", used);

      nest_out();
    }
    nest_out();
  }
}

void test_main(void) {
  test_average();
}

/* vim:ts=2:sw=2:sts=2:et:ft=c
 */

File: test/Downtown/Average.js
"use strict";

var chai = require("chai");
var expect = chai.expect;

var Average = require("../../lib/Downtown/Average.js");

function computeAverage(fn, from, to) {
  var total = 0;
  var min = NaN;
  var max = NaN;

  for (var i = from; i <= to; i++) {
    var datum = fn(i);
    if (isNaN(min) || datum < min) min = datum;
    if (isNaN(max) || datum > max) max = datum;
    total += datum;
  }

  return {
    min: min,
    max: max,
    avg: total / (to - from + 1)
  }
}

describe("Average", function() {

  function xn(n) {
    return ((100 - n) * 3.5 * ((n & 1) ? -1.3 : 3.2));
  }

  describe("Basic", function() {

    var a = new Average(5);

    it("should implement add, length & ready", function() {
      expect(Average).to.respondTo('add');
      expect(Average).to.respondTo('length');

      expect(a.length()).to.equal(0);
      expect(a.ready()).to.be.false;

      a.add();
      expect(a.length()).to.equal(0);
      expect(a.ready()).to.be.false;

      a.add(1);
      expect(a.length()).to.equal(1);
      expect(a.ready()).to.be.false;

      a.add(2, 4);
      expect(a.length()).to.equal(3);
      expect(a.ready()).to.be.true;
    });

    it("it should calculate moving average, min & max", function() {

      expect(a.average()).to.equal((1 + 2 + 4) / 3);
      expect(a.min()).to.equal(1);
      expect(a.max()).to.equal(4);

      for (var i = 1; i < 20; i *= 2) {
        for (var j = 0; j < 5; j++) a.add(i);
        expect(a.average()).to.equal(i);
        expect(a.min()).to.equal(i);
        expect(a.max()).to.equal(i);
      }

    });

  });

  describe("Soak test", function() {

    it("should return sensible values", function() {

      for (var len = 1; len < 5; len++) {
        var a = new Average(len);

        for (var x = 0; x < len * 2; x++) {
          var datum = xn(x);
          a.add(datum);
          var got_avg = a.average();
          var got_min = a.min();
          var got_max = a.max();

          var used = Math.min(x + 1, len);
          var want = computeAverage(xn, x - used + 1, x);

          expect(a.length()).to.equal(used);
          expect(got_avg).to.be.closeTo(want.avg, 0.00001);
          expect(got_min).to.be.closeTo(want.min, 0.00001);
          expect(got_max).to.be.closeTo(want.max, 0.00001);
        }
      }
    });
  });

});
